### DISCUSSION TITLE
How can broadcast a custom notification on specefic channel?

### DISCUSSION THREAD

**Discussion by JayAyAre** [filamentphp/filament#14076] [Help]
### Package

Notifications

### Package Version

v3.0.0

### How can we help you?

Hello, im trying to make a notification that scope pusher channel to a specific tenant, because i have multitenants on my project (admin, tenant), and also i have multi-envs (local, staging for testing on an server, and prod). My main goal with this is to get a kind of way to make an filament notification overwriting 'broadcast' method, with this, being capable to make notifications on local in the admin panel, and only see this notifications on local and in the admin panel.

right now, if i try to do a broadcast notification in that same way, i will see the same notification on local, and staging (prod no bc dont have the notification).

This is what i trying to do (Is not working):

1 - first, i have defined the channel:
```

Broadcast::channel('Notifications.App.Models.User.{id}.{currentPanel}.{tenantSimpleId}', function ($user) {
    return Auth::check();
});

Being currentPanel 'admin', 'tenant' or 'participant'. will be depending the result of the next TenantHelper's function:
    public static function getCurrentPanel(): ?string
    {
        if (!self::isCentralDomain()) {
            return 'tenant';
        } elseif (self::isAdminDomain()) {
            return 'admin';
        } elseif (self::isParticipantDomain()) {
            return 'participant';
        }
        return null;
    }
```

2 - Im entends from filament's notification the next one class:

```
use App\Events\UserBroadcastNotification;
use Filament\Notifications\Notification;
use Illuminate\Contracts\Auth\Authenticatable;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Support\Collection;

class UserScopedNotification extends Notification
{
    protected string $channelName;

    public function broadcast(Model | Authenticatable | Collection | array $users, ?string $channel = null): static
    {
        if (! is_iterable($users)) {
            $users = [$users];
        }

        foreach ($users as $user) {
            $currentPanel = $channel;
            $tenantSimpleId = tenant() ? '.' . tenant()->simple_id : '';
            $this->channelName = "Notifications.App.Models.User.{$user->getKey()}.{$currentPanel}{$tenantSimpleId}";
            $broadcastNotification = $this->toBroadcast();
            $user->notify($broadcastNotification);
        }

        return $this;
    }

    public function toBroadcast(): UserBroadcastNotification
    {
        $data = $this->toArray();
        $data['format'] = 'filament';
        return new UserBroadcastNotification($data, $this->channelName);
    }
}

With this, im trying to do the same static notfication to call diferents methods on the same instance. like this:

UserScopedNotification::make()
                            ->color('success')
                            ->icon('heroicon-o-check-circle')
                            ->title('PDF generated sucacessfully')
                            ->broadcast(auth()->user(), TenantHelper::getCurrentPanel()); // Getting the user and current panel (admin, tenant, participant....)

```
3- Then, i have defined the broadcast event with the same channel:

```
<?php

namespace App\Events;

use Filament\Notifications\BroadcastNotification;
use Illuminate\Broadcasting\InteractsWithSockets;
use Illuminate\Broadcasting\PrivateChannel;
use Illuminate\Contracts\Broadcasting\ShouldBroadcast;
use Illuminate\Foundation\Events\Dispatchable;
use Illuminate\Queue\SerializesModels;

class UserBroadcastNotification extends BroadcastNotification implements ShouldBroadcast
{
    use Dispatchable, InteractsWithSockets, SerializesModels;

    public function __construct(public array $data, public string $channelName)
    {
        parent::__construct($data);
    }

    public function broadcastOn(): array
    {
        return [new PrivateChannel($this->channelName)];
    }
}
```

4-  Finally, im trying to listen this event on a custom livewire component extending from Filament's notification componente:
```
use App\Helpers\TenantHelper;
use Filament\Notifications\Livewire\Notifications;

```
class LivewireBroadcastNotification extends Notifications
{
    /**
     * Get the broadcast channel for the authenticated user.
     * @return string|null
     */
    public function getBroadcastChannel(): ?string
    {
        $user = $this->getUser();

        if (! $user) {
            return null;
        }

        if (method_exists($user, 'receivesBroadcastNotificationsOn')) {
            return $user->receivesBroadcastNotificationsOn();
        }
        $userId = auth()->user()->getKey();
        $currentPanel = TenantHelper::getCurrentPanel();
        $tenantSimpleId = tenant() ? '.' . tenant()->simple_id : '';
        return "Notifications.App.Models.User.{$userId}.{$currentPanel}{$tenantSimpleId}";
    }
}
```

and i put this livewire like this on admin panel:

 public function boot(): void {
        $this->app->bind(Notification::class, UserScopedNotification::class);
    }


$panel...
->databaseNotifications()
            ->renderHook(
                PanelsRenderHook::BODY_END,
                fn(): string => Blade::render('@livewire(\'livewire-broadcast-notification\')'),
            )

```

**Comment by danihidayatx** [comment]
hello, did you find solutions for this?

**Comment by jjrohrer** [comment]
Nudge

**Comment by danihidayatx** [comment]
I eventually found a much simpler solution and it turned out I didn't need to override Filament's notification classes or Livewire component at all.

The key is making Laravel's broadcast notification channel tenant-aware.

On the notifiable model:

```php
public function receivesBroadcastNotificationsOn(): ?string
{
    if (! tenancy()->initialized) {
        return null;
    }

    return 'tenant.' . tenant('id') . '.user.' . $this->getKey() . '.notifications';
}
```

Then authorize the same channel:

```php
Broadcast::channel(
    'tenant.{tenant}.user.{id}.notifications',
    fn ($user, $tenant, $id) => (int) $user->id === (int) $id,
);
```

**Also make sure `/broadcasting/auth` is executed inside the tenant context (tenant initialization before channel authorization).**

Once those pieces are in place, Filament automatically subscribes to the correct Echo channel because it already calls `receivesBroadcastNotificationsOn()` internally.

After that, the standard API works without any custom notification classes:

```php
Notification::make()
    ->title('Hello!')
    ->broadcast($user);
```

So the issue wasn't really Filament—it was that the broadcast channel wasn't tenant-aware.

Hopefully this helps anyone trying to make Filament notifications work with Stancl Tenancy or a similar multi-tenant setup.
